libursa/src/signatures/bls.rs

```diff
@@ -0,0 +1,706 @@
+                assert!(!sig.verify_multi(inputs.as_slice(), &g));
+            }
+        }
+    };
+}
+
+pub mod prelude {
+    pub use super::{small::*, PrivateKey};
+}
+
+/// This version is the small BLS signature scheme
+/// with the public key group in G1 and signature group in G2.
+/// 192 byte signatures and 97 byte public keys
+pub mod normal {
+    use super::*;
+
+    pub const PUBLIC_KEY_SIZE: usize = GroupG1_SIZE;
+    pub const SIGNATURE_SIZE: usize = GroupG2_SIZE;
+
+    pub type Generator = G1;
+    pub type SignatureGroup = G2;
+
+    public_key_impl!();
+
+    aggregate_public_key_impl!();
+
+    #[derive(Debug, Clone, Serialize, Deserialize)]
+    pub struct Signature(SignatureGroup);
+
+    impl Signature {
+        pub fn new(message: &[u8], sk: &PrivateKey) -> Self {
+            Signature(&SignatureGroup::from_msg_hash(message) * sk)
+        }
+
+        pub fn new_with_rk_mitigation(
+            message: &[u8],
+            sk: &PrivateKey,
+            pk_index: usize,
+            pks: &[PublicKey],
+        ) -> Self {
+            // To combat the rogue key attack,
+            // compute (t_1,…,t_n)←H1(pk_1,…,pk_n) ∈ R_n
+            // output the aggregated public key
+            // as described in section 3.1 from https://eprint.iacr.org/2018/483
+            let mut bytes = Vec::new();
+            for k in pks {
+                bytes.extend_from_slice(k.to_bytes().as_slice());
+            }
+            bytes.extend_from_slice(pks[pk_index].to_bytes().as_slice());
+            let a = FieldElement::from_msg_hash(bytes.as_slice());
+            Signature(SignatureGroup::from_msg_hash(message) * sk * &a)
+        }
+
+        // Collects multiple signatures into a single signature
+        // Verified by using `verify_multi`. This method does not
+        // directly mitigate the rogue key attack. It is expected the caller
+        // handles this using other techniques like proof of possession
+        pub fn combine(&mut self, signatures: &[Signature]) {
+            for sig in signatures {
+                self.0 += &sig.0;
+            }
+        }
+
+        // Verify a signature generated by `new`
+        pub fn verify(&self, message: &[u8], pk: &PublicKey, g: &Generator) -> bool {
+            let hash = SignatureGroup::from_msg_hash(message);
+            GT::ate_2_pairing(&-g, &self.0, &pk.0, &hash).is_one()
+        }
+
+        // Caller should aggregate all signatures into `self` by using `combine`.
+        // Messages must be distinct
+        // `inputs` is a slice of message - public key tuples
+        // Multisignature verification
+        pub fn verify_multi(&self, inputs: &[(&[u8], &PublicKey)], g: &Generator) -> bool {
+            let mut msg_check = ::std::collections::HashSet::new();
+            let mut pairs = Vec::new();
+            for (msg, pk) in inputs {
+                let hash = SignatureGroup::from_msg_hash(&msg);
```

Comment:
Domain separation?

Reply:
@lovesh, The caller is the one that knows what is being signed unless we have a function for each specific purpose. The general practice is to prepend or append bytes to the message being signed. If we want to be able to select a different XOF depending on the type of data being signed we should expose that. But in general, signing libraries do not handle domain separation directly. OpenSSL, other SSL variants, Libsodium, libsecp256k1 all do not. Dalek crypto offers a byte array as an optional parameter [here](https://github.com/dalek-cryptography/ed25519-dalek/blob/5870607aa770f8f08114fb8a8eae7e605b82b7c9/src/secret.rs#L482) but doesn't enforce it. Why?  Because it's up to the caller to know what they are signing. The library just supplies the methods to sign. The library has no idea what is being signed unless it has a specific function called `sign_public_key` or `sign_general_message`. In those cases, the only difference between them is the XOF  to hash to point or a fixed prefix. I believe the caller would like to have these defined for their system parameters. Fundamentally they are exactly the same in every way minus that detail. Having two methods like that is okay but the majority of the time, only the one method is used. So can you give me a good reason why we should add domain separation into the library vs punting to the user to add their own domain separation?

Reply:
How is XOF relevant here? 
**As i said in my previous comment, the separation of domains is done between messages and verkeys.** 
All messages being signed will have the same prefix prepended/appended before hashing them. This prefix will be different than the prefix prepended/appended before hashing public keys but public keys will have the same prefix. The reason for this separation is to get separate random oracles (as the paper demands it and will be needed for security proof) built from the same hash function. 

> So can you give me a good reason why we should add domain separation into the library vs punting to the user to add their own domain separation?

Because a domain separator chosen by the user will not have any special property than a separator chosen by the implementation. It will complicate the interface and then puts the burnden on the caller(s) to use differen separators for messages and verkeys.